avro-fastserde/src/main/java/com/linkedin/avro/fastserde/PrimitiveFloatList.java

```diff
@@ -60,44 +75,72 @@ public PrimitiveFloatList(Collection<Float> c) {
    * @throws IOException on io errors
    */
   public static Object readPrimitiveFloatArray(Object old, Decoder in) throws IOException {
-    long l = in.readArrayStart();
-    if (l > 0) {
-      PrimitiveFloatList array = (PrimitiveFloatList) newPrimitiveFloatArray(old, (int) l);
+    long length = in.readArrayStart();
+    long totalLenth = 0;
+
+    if (length > 0) {
+      PrimitiveFloatList array = (PrimitiveFloatList) newPrimitiveFloatArray(old);
+
       do {
-        for (long i = 0; i < l; i++) {
-          array.addPrimitive(in.readFloat());
+        // Allocate ByeBuffer of size array length * Float.BYTES to hold the array values
+        ByteBuffer byteBuffer = ByteBuffer.allocate((int)length*4).order(ByteOrder.LITTLE_ENDIAN);
+        ByteBufferBinaryDecoder binaryDecoder;
+        if (in instanceof ByteBufferBinaryDecoder) {
+          binaryDecoder = (ByteBufferBinaryDecoder)in;
+          binaryDecoder.readBytes(byteBuffer.array(),0, (int) (length << 2));
+        } else {
+          binaryDecoder = new ByteBufferBinaryDecoder((BinaryDecoder)in);
+          binaryDecoder.readBytes(byteBuffer.array(), (int) (length << 2));
+          // skip array bytes + array end
+          in.skipFixed(((int)length << 2) + 1);
         }
-        l = in.arrayNext();
-      } while (l > 0);
+        array.byteBuffers.add(byteBuffer);
+        totalLenth += length;
+        length = binaryDecoder.arrayNext();
+      } while (length > 0);
+
+      setupElements(old, array, (int)totalLenth);
       return array;
     } else {
-      return newPrimitiveFloatArray(old, 0);
+      return new PrimitiveFloatList(0);
```

Comment:
is this immutable? if so can this be made a constant? (like COllections.emptyList())

Reply:
Probably not do it as for reuse case we could reuse it.